guidelines/auth.md

```diff
@@ -0,0 +1,126 @@
+struct that Kubernetes uses to hold the user’s identity. 
+
+#### Recommendation
+
+Every web app should accept the following headers for user identity:
+* A username header, configurable via the `KUBEFLOW_USERID_HEADER` setting 
+    * Some IdPs (like Google) add a prefix to this header (e.g.,
+      `accounts.google.com:`). If the Web App wants to support stripping that
+      prefix, it should expose a `KUBEFLOW_USERID_PREFIX` setting).
+* A group header, configurable via the `KUBEFLOW_GROUPS_HEADER` setting.
+* A prefix for headers containing extra identity information (e.g., scopes),
+  configurable via the `KUBEFLOW_EXTRAS_HEADER_PREFIX` setting. The prefix is
+  removed, the remainder header name is percent decoded and becomes the key, the
+  header value becomes the value. ([see also](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy)).
+
+
+### Authorization
+
+For authorization, we want to still consult the K8s RBAC permissions. To do
+that, there are two ways:
+* [SubjectAccessReview API](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#checking-api-access),
+  allows the Web App to ask the Kubernetes RBAC authorizer if an action is
+  allowed by a specific identity (user, group, etc.). After the check passes,
+  the request is made with the Web App’s credentials.
+    * Minimal addition of permissions to Web Apps (only permissions to perform
+      SubjectAccessReview). 
+    * API-Server logs show the web-app as the performer of the API calls.
+      SubjectAccessReview calls and subsequent API calls could be correlated.
+    * Requires 2 requests and doesn’t catch some cases ( RBAC privilege
+      escalation checks, PSPs, Admission Controllers that rely on the user’s
+      identity).
+* [Impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation)
+  allows the Web App to perform the request to the K8s API Server as if it was
+  made from a specified identity (user, group, etc.). Here, the check is made by
+  the K8s API Server on the impersonated identity.
+    * Requires only 1 request and passes through all authorization checks of
+      K8s, some of which aren’t caught by SubjectAccessReview ( RBAC privilege
+      escalation checks, PSPs, Admission Controllers that rely on the user’s
+      identity).
+    * The issue with impersonation is that it is too powerful and there is no
+      way to scope it down. For example, if the Jupyter Notebooks Web App used
+      impersonation, it would have the full permissions of every user it could
+      impersonate. This presents too much of a surface for Web Apps that would
+      normally need little permissions, like the Jupyter Notebook Web App. For
+      Web Apps that would require elevated permissions anyway (e.g. an admin
+      panel web app), impersonation could be a viable option.
+
```

Comment:
How about adding a reference section with reference to past docs, discussions, etc...?

Reply:
Sounds good, I'll work on such a section.